ui-trackoccupancydiagram/src/components/helpers/drawElements/drawOccupancyZones.ts

```diff
@@ -4,58 +4,202 @@ import {
+  ctx.lineTo(arrivalTime + ARROW_OFFSET_X, OCCUPANCY_ZONE_START + ARROW_OFFSET_Y);
+  ctx.stroke();
+};
+
+type DrawRemainingTrainsBox = {
+  ctx: CanvasRenderingContext2D;
+  text: string;
+  textX: number;
+  textY: number;
+};
+
+const drawRemainingTrainsBox = ({ ctx, text, textX, textY }: DrawRemainingTrainsBox) => {
+  ctx.fillStyle = REMAINING_TRAINS_BACKGROUND;
+  ctx.beginPath();
+  ctx.rect(textX, textY, REMAINING_TRAINS_WIDTH, REMAINING_TRAINS_HEIGHT);
+  ctx.fill();
+  ctx.stroke();
+  ctx.fillStyle = WHITE_100;
+  ctx.font = SANS;
+  ctx.textAlign = 'center';
+  ctx.textBaseline = 'middle';
+  ctx.fillText(text, textX + REMAINING_TRAINS_WIDTH / 2, textY + REMAINING_TRAINS_HEIGHT / 2);
+};
+
 export const drawOccupancyZones = ({
   ctx,
   width,
   height,
   tracks,
   occupancyZones,
   getTimePixel,
+  selectedTrainId,
 }: {
   ctx: CanvasRenderingContext2D;
   width: number;
   height: number;
   tracks: Track[] | undefined;
   occupancyZones: OccupancyZone[] | undefined;
   getTimePixel: (time: number) => number;
+  selectedTrainId: string;
 }) => {
   ctx.clearRect(0, 0, width, height);
 
   tracks?.forEach((track, index) => {
     const trackTranslate = index === 0 ? CANVAS_PADDING : TRACK_HEIGHT_CONTAINER;
     ctx.translate(0, trackTranslate);
 
-    const trackOccupancyZones = occupancyZones?.filter((zone) => zone.trackId === track.id);
+    const filteredOccupancyZone = occupancyZones?.filter((zone) => zone.trackId === track.id);
+    const sortedOccupancyZone = filteredOccupancyZone?.sort(
+      (a, b) => a.arrivalTime.getTime() - b.arrivalTime.getTime()
+    );
+
+    if (!sortedOccupancyZone) return;
 
-    if (!trackOccupancyZones) return;
+    let primaryArrivalTimePixel = 0;
+    let primaryDepartureTimePixel = 0;
+    let lastDepartureTimePixel = 0;
+    let yPosition = OCCUPANCY_ZONE_START;
+    let yOffset = 0;
+    let counter = 0;
 
-    trackOccupancyZones.forEach((zone) => {
+    if (sortedOccupancyZone && sortedOccupancyZone.length > 1) {
+      primaryArrivalTimePixel = getTimePixel(sortedOccupancyZone[0].arrivalTime.getTime());
+      primaryDepartureTimePixel = getTimePixel(sortedOccupancyZone[0].departureTime.getTime());
+      lastDepartureTimePixel = primaryDepartureTimePixel;
+    }
+
+    // use a for of loop to be able to break the loop if there are more than 9 zones
+    for (const [sortedIndex, zone] of sortedOccupancyZone.entries()) {
       const arrivalTime = getTimePixel(zone.arrivalTime.getTime());
       const departureTime = getTimePixel(zone.departureTime.getTime());
+      const isThroughTrain = arrivalTime === departureTime;
+
+      // reset the overlap check if the zone is not overlapping
+      if (arrivalTime > lastDepartureTimePixel) {
+        yPosition = OCCUPANCY_ZONE_START;
+        primaryArrivalTimePixel = arrivalTime;
+        primaryDepartureTimePixel = departureTime;
+        lastDepartureTimePixel = departureTime;
+        yOffset = 0;
+        counter = 0;
+      }
```

Comment:
I think it would be nice to draw the occupationZone here and then continue. This way, we don't have to handle the other case.

You can extract the code between lines 155-190.

```suggestion
      if (arrivalTime > lastDepartureTimePixel) {
        yPosition = OCCUPANCY_ZONE_START;
        primaryArrivalTimePixel = arrivalTime;
        primaryDepartureTimePixel = departureTime;
        lastDepartureTimePixel = departureTime;
        yOffset = 0;
        counter = 0;
        
        drawOccupationZone(...);
        continue;
      }
```

Reply:
if I do that, I have to remove the `else if`, `drawRemainingTrainsBox()` will be cast at each iteration which is bad, then we won't see the opacity of the background. `break` gives me the assurance of stopping any action in the loop, then draw the remaining trains at last

Reply:
fixed by using a while loop as we talk earlier

Reply:
Hum I don't really get your point.

I asked you this change to simplify the logic of this function, which would be smth like this:
```
if zone is not overlapping:
  then draw the zone and handle the next zone
else:
  then handle overlapping zone
```

With the current logic, you mixed the 2 cases (overlapping and not overlapping zones) till l211, which is not necessary

Reply:
my bad, I didn't get it this way, it seemed to me enough better to fix the `break` issue and still have a way to increment on the index generated, now it's separated